snakefmt/shell_formatter.py

```
@@ -0,0 +1,153 @@
+    """Run the bundled shfmt binary against `code`. Seam for future swaps —
+    see docs/adr/0001-shell-formatter-distribution.md."""
+    try:
+        process = subprocess.run(
+            ["shfmt", "-i", "4", "-ci", "-bn"],
+            input=code,
+            text=True,
+            capture_output=True,
+            check=True,
+        )
+    except subprocess.CalledProcessError as e:
+        raise InvalidShell(f"Invalid shell code. shfmt error: {e.stderr}")
```

Ruff (0.15.12)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)